File: Sources/RefId/Encoding.swift
// SPDX-License-Identifier: Apache-2.0
//
// Percent-encode/decode helpers driven by an `encoding.<name>.table` from the spec — never a
// hardcoded character list. Which table applies to a form is the form's own `encoding` field.

enum Encoding {
    /// The encoding table a form declares, or none when the form declares no encoding.
    static func table(_ spec: Spec, form: [String: Any]) -> [String: String] {
        guard let name = form["encoding"] as? String else { return [:] }
        return spec.table("encoding", name, "table")
    }

    static func containsAny(_ raw: String, _ characters: [Character]) -> Bool {
        raw.contains { characters.contains($0) }
    }

    /// One left-to-right pass over the source characters; a produced percent-form is never re-scanned.
    static func encode(_ raw: String, table: [String: String]) -> String {
        if table.isEmpty { return raw }
        var out = ""
        for character in raw {
            out += table[String(character)] ?? String(character)
        }
        return out
    }

    /// One left-to-right pass over the encoded text; decoding `%2523` yields `%23`, never `#`.
    static func decode(_ encoded: String, table: [String: String]) -> String {
        if table.isEmpty { return encoded }
        let reverse = Dictionary(uniqueKeysWithValues: table.map { ($1, $0) })
        let forms = Array(reverse.keys)
        var out = ""
        var index = encoded.startIndex
        while index < encoded.endIndex {
            if let form = forms.first(where: { encoded[index...].hasPrefix($0) }) {
                out += reverse[form] ?? form
                index = encoded.index(index, offsetBy: form.count)
            } else {
                out.append(encoded[index])
                index = encoded.index(after: index)
            }
        }
        return out
    }

    /// True when every `%` in the value begins one of the table's percent-forms — the strict nested encoding.
    static func strictlyEncoded(_ value: String, table: [String: String]) -> Bool {
        let forms = Array(table.values)
        var index = value.startIndex
        while let at = value[index...].firstIndex(of: "%") {
            guard forms.contains(where: { value[at...].hasPrefix($0) }) else { return false }
            index = value.index(after: at)
        }
        return true
    }
}

File: crates/ref-id/src/encoding.rs
// SPDX-License-Identifier: Apache-2.0
//
// Percent-encode/decode helpers driven by an `encoding.<name>.table` from the spec — never a
// hardcoded character list. Which table applies to a form is the form's own `encoding` field.

use crate::spec::Spec;
use serde_json::Value;

/// The encoding table a form declares, or none when the form declares no encoding.
pub(crate) fn table_for(spec: &Spec, form: &Value) -> Vec<(String, String)> {
    match form.get("encoding").and_then(Value::as_str) {
        Some(name) => spec.table(&["encoding", name, "table"]),
        None => Vec::new(),
    }
}

pub(crate) fn contains_any(raw: &str, characters: &[char]) -> bool {
    raw.chars().any(|c| characters.contains(&c))
}

/// One left-to-right pass over the source characters; a produced percent-form is never re-scanned.
pub(crate) fn encode(raw: &str, table: &[(String, String)]) -> String {
    if table.is_empty() {
        return raw.to_string();
    }
    let mut out = String::with_capacity(raw.len());
    for c in raw.chars() {
        match table.iter().find(|(k, _)| k.chars().next() == Some(c) && k.chars().count() == 1) {
            Some((_, form)) => out.push_str(form),
            None => out.push(c),
        }
    }
    out
}

/// One left-to-right pass over the encoded text; decoding `%2523` yields `%23`, never `#`.
pub(crate) fn decode(encoded: &str, table: &[(String, String)]) -> String {
    if table.is_empty() {
        return encoded.to_string();
    }
    let mut out = String::with_capacity(encoded.len());
    let mut rest = encoded;
    while !rest.is_empty() {
        if let Some((character, form)) = table.iter().find(|(_, form)| rest.starts_with(form.as_str())) {
            out.push_str(character);
            rest = &rest[form.len()..];
        } else {
            let c = rest.chars().next().unwrap();
            out.push(c);
            rest = &rest[c.len_utf8()..];
        }
    }
    out
}

/// True when every `%` in the value begins one of the table's percent-forms — the strict nested encoding.
pub(crate) fn strictly_encoded(value: &str, table: &[(String, String)]) -> bool {
    value.match_indices('%').all(|(at, _)| table.iter().any(|(_, form)| value[at..].starts_with(form.as_str())))
}
